socs/agents/ld_monitor/agent.py

```diff
@@ -0,0 +1,425 @@
+        Defines verbosity for debugging purposes
+    host : string
+        Defines the host where data will be received (where the agent is to be ran)
+    port : int
+        Port number in the local host to be bound to receive the data
+    sockopen : bool
+        Indicates when the socket is open
+    inittime : float
+        Logs the time at which initialization was carried out
+    data_dict : dictionary
+        Raw data received from the lightning detector
+    newdata_dict : dictioanry
+        The dictionary where new data is received
+    """
+
+    def __init__(self, port=1110, verbose=verbosity):
+        self.verbose = verbose
+        self.port = port
+
+        # get localhost ip
+        with socket.socket(socket.AF_INET, socket.SOCK_DGRAM) as s:
+            s.connect(("8.8.8.8", 80))
+            self.host = s.getsockname()[0]
+
+        if hasattr(self, 'sockopen'):
+            self.sock.close()
+
+        # open and bind socket to receive lightning detector data
+        try:
+            self.sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
+        except BaseException:
+            print('Failed to create socket')
+
+        try:
+            self.sock.bind((self.host, self.port))
+            self.sockopen = True
+            self.inittime = time.time()
+        except BaseException:
+            print('Failed to bind socket')
+
+        # initialize variables to account for absence of previous data
+        self.data_dict = {
+            'd_type': numpy.nan,
+            'field_value': numpy.nan,
+            'rot_fault': 0,
+            'time_last': -1.,
+            'tsince_last': -1.,
+            'dist': -1,
+            'unit_d': 0,
+            'high_field': -1,
+            'hifield_value': -1000.,
+            'alarm_r': 0,
+            'alarm_o': 0,
+            'alarm_y': 0,
+            'delay_g': 1,
+            'clear': 0,
+            'r_timer': 0,
+            'o_timer': 0,
+            'y_timer': 0,
+            'g_timer': 0,
+            'allclear_timer': 0,
+            'faultcode': 0
+        }
+
+        if self.verbose:
+            print('ld_monitor function monitor initialized')
+
+    def read_data(self):
+        """
+        Receives data from the lightning detector via UDP,
+        and returns a dictionary containing formatted data
+        """
+
+        self.data, _ = self.sock.recvfrom(1024)
+        self.data = self.data.decode('utf-8')
+
+        # receiving an "e-fiel" sentence
+        if self.data[0] == '$':
+            data_split = self.data[1:].split(',')
+            rot_fault = int(data_split[1].split('*')[0])
+            self.newdata_dict = {
+                'd_type': 0,
+                'field_value': float(data_split[0]),
+                'rot_fault': rot_fault
+            }
+            self.data_dict.update(self.newdata_dict)
+            return self.data_dict
+
+        elif self.data[0] == '@':
+            param = self.data[1:3]
+
+            match param:

```

Comment:
randomly saw this. This only works on Python>=3.10; SO requires code to work with Python >=3.8